File: lib/models/order.dart
import 'cart_item.dart';

class Order {
  final int id;
  final int userId;
  final List<CartItem> items;
  final double subtotal;
  final double shipping;
  final double tax;
  final double total;
  final String status;
  final String? shippingAddress;
  final String? paymentMethod;
  final String? notes;
  final DateTime createdAt;
  final DateTime? updatedAt;

  Order({
    required this.id,
    required this.userId,
    this.items = const [],
    required this.subtotal,
    this.shipping = 0,
    this.tax = 0,
    required this.total,
    required this.status,
    this.shippingAddress,
    this.paymentMethod,
    this.notes,
    required this.createdAt,
    this.updatedAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    List<CartItem> orderItems = [];
    if (json['items'] != null && json['items'] is List) {
      orderItems = (json['items'] as List)
          .map((item) => CartItem.fromJson(item))
          .toList();
    }

    return Order(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      items: orderItems,
      subtotal: (json['subtotal'] ?? 0).toDouble(),
      shipping: (json['shipping'] ?? 0).toDouble(),
      tax: (json['tax'] ?? 0).toDouble(),
      total: (json['total'] ?? 0).toDouble(),
      status: json['status'] ?? 'pending',
      shippingAddress: json['shipping_address'],
      paymentMethod: json['payment_method'],
      notes: json['notes'],
      createdAt: json['created_at'] != null 
          ? DateTime.parse(json['created_at']) 
          : DateTime.now(),
      updatedAt: json['updated_at'] != null 
          ? DateTime.parse(json['updated_at']) 
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': userId,
      'items': items.map((item) => item.toJson()).toList(),
      'subtotal': subtotal,
      'shipping': shipping,
      'tax': tax,
      'total': total,
      'status': status,
      'shipping_address': shippingAddress,
      'payment_method': paymentMethod,
      'notes': notes,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt?.toIso8601String(),
    };
  }

  String get statusText {
    switch (status) {
      case 'pending':
        return 'قيد الانتظار';
      case 'processing':
        return 'قيد المعالجة';
      case 'shipped':
        return 'تم الشحن';
      case 'delivered':
        return 'تم التوصيل';
      case 'cancelled':
        return 'ملغي';
      default:
        return status;
    }
  }
}

File: php-project/app/Models/Order.php
<?php

namespace App\Models;

use App\Database\Connection;

class Order
{
    private string $table = 'orders';
    private \PDO $db;
    
    public function __construct() { $this->db = Connection::getInstance()->getConnection(); }
    
    public function getByUser(int $userId, int $page = 1, int $perPage = 10): array
    {
        $offset = ($page - 1) * $perPage;
        $countStmt = $this->db->prepare("SELECT COUNT(*) FROM {$this->table} WHERE user_id = ?"); $countStmt->execute([$userId]);
        $stmt = $this->db->prepare("SELECT * FROM {$this->table} WHERE user_id = ? ORDER BY created_at DESC LIMIT ? OFFSET ?"); $stmt->execute([$userId, $perPage, $offset]);
        return ['orders' => $stmt->fetchAll(), 'total' => $countStmt->fetchColumn()];
    }
    
    public function findById(int $id): ?array { $stmt = $this->db->prepare("SELECT * FROM {$this->table} WHERE id = ?"); $stmt->execute([$id]); $order = $stmt->fetch(); if ($order) { $order['items'] = $this->getItems($id); } return $order ?: null; }
    
    public function getItems(int $orderId): array { $stmt = $this->db->prepare("SELECT oi.*, p.name, p.image FROM order_items oi JOIN products p ON oi.product_id = p.id WHERE oi.order_id = ?"); $stmt->execute([$orderId]); return $stmt->fetchAll(); }
    
    public function create(int $userId, array $data): array
    {
        $cart = (new Cart())->getByUser($userId);
        if (empty($cart['items'])) return ['success' => false, 'message' => 'السلة فارغة'];
        $totals = (new Cart())->calculateTotals($userId, $data['coupon_code'] ?? null);
        $orderNumber = 'ORD-' . date('Ymd') . '-' . strtoupper(substr(uniqid(), -6));
        $this->db->prepare("INSERT INTO {$this->table} (user_id, order_number, subtotal, discount, tax, shipping, total, status, shipping_address, payment_method) VALUES (?, ?, ?, ?, ?, ?, ?, 'pending', ?, ?)")->execute([$userId, $orderNumber, $totals['subtotal'], $totals['discount'], $totals['tax'], $totals['shipping'], $totals['total'], json_encode($data['shipping_address'] ?? []), $data['payment_method'] ?? 'cod']);
        $orderId = (int) $this->db->lastInsertId();
        foreach ($cart['items'] as $item) { $this->db->prepare("INSERT INTO order_items (order_id, product_id, variant_id, quantity, price, total) VALUES (?, ?, ?, ?, ?, ?)")->execute([$orderId, $item['product_id'], $item['variant_id'] ?? null, $item['quantity'], $item['unit_price'], $item['total']]); }
        (new Cart())->clear($userId);
        return ['success' => true, 'order' => $this->findById($orderId)];
    }
    
    public function cancel(int $id, int $userId): array
    {
        $order = $this->findById($id);
        if (!$order || $order['user_id'] != $userId) return ['success' => false, 'message' => 'الطلب غير موجود'];
        if (!in_array($order['status'], ['pending', 'processing'])) return ['success' => false, 'message' => 'لا يمكن إلغاء هذا الطلب'];
        $this->db->prepare("UPDATE {$this->table} SET status = 'cancelled' WHERE id = ?")->execute([$id]);
        return ['success' => true, 'order' => $this->findById($id)];
    }
}
